src/middle/SizedType.ml

```diff
@@ -220,6 +220,15 @@ let modify_sizedtype_mem (mem_pattern : Mem_pattern.t) st =
   match mem_pattern with
   | AoS -> demote_sizedtype_mem st
   | SoA -> promote_sizedtype_mem st
+  | OpenCL -> promote_sizedtype_mem st
+
+let rec promote_mem (mem_pattern : Mem_pattern.t) st =
+  match st with
+  | SVector (_, dim) -> SVector (mem_pattern, dim)
+  | SRowVector (_, dim) -> SRowVector (mem_pattern, dim)
+  | SMatrix (_, dim1, dim2) -> SMatrix (mem_pattern, dim1, dim2)
+  | SArray (inner_type, dim) -> SArray (promote_mem mem_pattern inner_type, dim)
+  | _ -> st
```

Comment:
Does this need to consider things inside tuples?

Reply:
If we want to allow OpenCL to use tuples then yes. 

Also the `Decls` having a mem_pattern tag won't work because of tuples either :( I think we do just need to tag all sized types as having a memory pattern

Reply:
Ah, that’s too bad. You could do something like what we had to do for Autodiff level and have a tuple specific variant in the type, but I wasn’t super happy with that either.

Reply:
Yeah eod I think it's going to look a little odd in some places but I think it's fine to just add it to the sized types